lib/sidekiq/throttled/expirable_set.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+require "concurrent"
+
+module Sidekiq
+  module Throttled
+    # @api internal
+    #
+    # Set of elements with expirations.
+    #
+    # @example
+    #   set = ExpirableSet.new(10.0)
+    #   set.add("a")
+    #   sleep(5)
+    #   set.add("b")
+    #   set.to_a # => ["a", "b"]
+    #   sleep(5)
+    #   set.to_a # => ["b"]
+    class ExpirableSet
+      include Enumerable
+
+      # @param ttl [Float] expiration is seconds
+      # @raise [ArgumentError] if `ttl` is not positive Float
+      def initialize(ttl)
+        raise ArgumentError, "ttl must be positive Float" unless ttl.is_a?(Float) && ttl.positive?
+
+        @elements = Concurrent::Map.new
+        @ttl      = ttl
+      end
+
+      # @param element [Object]
+      # @return [ExpirableSet] self
+      def add(element)
+        @elements[element] = now + @ttl
+        self
+      end
+
+      # @yield [Object] Gives each live (not expired) element to the block
+      def each
+        return to_enum __method__ unless block_given?
+
+        horizon = now
+
+        @elements.each_pair do |element, sunset|
+          yield element if horizon < sunset
+        end
```

Comment:
It might be a source of memory leaks if expired keys will not be deleted over time. I propose to delete the expired elements from the list:

```suggestion
        @elements.keep_if do |element, sunset|
          horizon >= sunset
        end.each_key do |element|
          yield element
        end
```

Reply:
`Concurrent::Map` does not have `keep_if`, but I agree with idea. Will fix it.

Reply:
Addressed in: 6fb5e91